src/Data/Serialize/Get.hs

```diff
@@ -402,30 +418,101 @@ ensure n0 = n0 `seq` Get $ \ s0 b0 m0 w0 kf ks -> let
                     in ks s b m0 w0 s
                 else getMore n' s0 ss b0 m0 w0 kf ks
 
+negativeIsolation :: Get a
+negativeIsolation = fail "Attempted to isolate a negative number of bytes"
+
+isolationUnderParse :: Get a
+isolationUnderParse =  fail "Isolated parser didn't consume all input"
+
+isolationUnderSupply :: Get a
+isolationUnderSupply = fail "Too few bytes supplied to isolated parser"
+
+isolate0 :: Get a -> Get a
+isolate0 parser = do
+  rest <- get
+  cur <- bytesRead
+  put mempty cur
+  a    <- parser
+  put rest cur
+  pure a
+
 -- | Isolate an action to operating within a fixed block of bytes.  The action
 --   is required to consume all the bytes that it is isolated to.
 isolate :: Int -> Get a -> Get a
-isolate n m = do
-  M.when (n < 0) (fail "Attempted to isolate a negative number of bytes")
-  s <- ensure n
-  let (s',rest) = B.splitAt n s
-  cur <- bytesRead
-  put s' cur
-  a    <- m
-  used <- get
-  unless (B.null used) (fail "not all bytes parsed in isolate")
-  put rest (cur + n)
-  return a
+isolate n m
+  | n < 0 = negativeIsolation
+  | n == 0 = isolate0 m
```

Comment:
Let's add a test case that shows that the `0` case works if there isn't already one. It'd be nice to be able to remove this guard and show that the test would have failed, proving this case is useful, and having the test will cover this code path.

Reply:
Okay, I've added some cases for `ensure` and `isolate` calls for zero bytes. Looks like I was overly cautious with my guards :)